psycopg/pqpath.c

```diff
@@ -1514,6 +1529,280 @@ _pq_copy_out_v3(cursorObject *curs)
+    }
+    else if (buffer[0] == 'k') {
+        /* msgtype(1), walEnd(8), sendTime(8), reply(1) */
+        hdr = 1 + 8 + 8;
+        if (len < hdr + 1) {
+            psyco_set_error(OperationalError, curs, "keepalive message header too small");
+            goto exit;
+        }
+
+        reply = buffer[hdr];
+        if (reply) {
+            if (!pq_send_replication_feedback(curs, 0)) {
+                if (curs->conn->async) {
+                    curs->repl_feedback_pending = 1;
+                } else {
+                    pq_raise(curs->conn, curs, NULL);
+                    goto exit;
+                }
+            }
+            else {
+                gettimeofday(&curs->repl_last_io, NULL);
+            }
+        }
+
+        PQfreemem(buffer);
+        buffer = NULL;
+        goto retry;
+    }
+    else {
+        psyco_set_error(OperationalError, curs, "unrecognized replication message type");
+        goto exit;
+    }
+
+exit:
+    if (buffer) {
+        PQfreemem(buffer);
+    }
+
+    return msg;
+
+none:
+    msg = Py_None;
+    Py_INCREF(msg);
+    goto exit;
+}
+
+int
+pq_send_replication_feedback(cursorObject* curs, int reply_requested)
+{
+    char replybuf[1 + 8 + 8 + 8 + 8 + 1];
+    int len = 0;
+
+    Dprintf("pq_send_replication_feedback: write="XLOGFMTSTR", flush="XLOGFMTSTR", apply="XLOGFMTSTR,
+            XLOGFMTARGS(curs->repl_write_lsn),
+            XLOGFMTARGS(curs->repl_flush_lsn),
+            XLOGFMTARGS(curs->repl_apply_lsn));
+
+    replybuf[len] = 'r'; len += 1;
+    fe_sendint64(curs->repl_write_lsn, &replybuf[len]); len += 8;
+    fe_sendint64(curs->repl_flush_lsn, &replybuf[len]); len += 8;
+    fe_sendint64(curs->repl_apply_lsn, &replybuf[len]); len += 8;
+    fe_sendint64(feGetCurrentTimestamp(), &replybuf[len]); len += 8;
+    replybuf[len] = reply_requested ? 1 : 0; len += 1;
+
+    if (PQputCopyData(curs->conn->pgconn, replybuf, len) <= 0 ||
+        PQflush(curs->conn->pgconn) != 0) {
+        return 0;
+    }
+    gettimeofday(&curs->repl_last_io, NULL);
+
+    return 1;
+}
+
+/* used for streaming replication only */
+static int
+_pq_copy_both_v3(cursorObject *curs)
+{
+    PyObject *msg, *tmp = NULL;
+    PyObject *write_func = NULL;
+    int is_text, fd, sel, ret = -1;
+    PGconn *pgconn;
+    fd_set fds;
+    struct timeval curr_time, ping_time, time_diff;
+
+    if (!curs->copyfile) {
+        psyco_set_error(ProgrammingError, curs,
+            "can't execute START_REPLICATION directly: use the start_replication() method instead");
+        goto exit;
+    }
+
+    if (!(write_func = PyObject_GetAttrString(curs->copyfile, "write"))) {
+        Dprintf("_pq_copy_both_v3: can't get o.write");
+        goto exit;
+    }
+
+    /* if the file is text we must pass it unicode. */
+    if (-1 == (is_text = psycopg_is_text_file(curs->copyfile))) {
+        goto exit;
+    }
+
+    CLEARPGRES(curs->pgres);
+    pgconn = curs->conn->pgconn;
+
+    while (1) {
```

Comment:
/*
 * This is synchronous mode ReplicationCursor.start_replication(...) loop, which consumes messages
 * and sends feedback and keepalives until asked to stop, feeding them to a user-provided callback
 * object.
 */